### PR TITLE
Allow setting replace patchStrategy for structs

### LINKED ISSUE #20208: strategicpatch.CreateTwoWayMergePatch cannot handle RawExtension fields
Extracted from https://github.com/kubernetes/kubernetes/pull/20093#issuecomment-175490502

The RawExtension type allows an API object to inline arbitrary JSON. strategicpatch.CreateTwoWayMergePatch returns an error trying to map that inlined JSON to struct fields. It should be able to compute a patch for an object containing a field of that type.

### PR DESCRIPTION
**Pull request by soltysh**
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #20208

@liggitt and @ymqytw ptal, you were in the original issue

### CODE REVIEW

**Comment by liggitt** [comment]
cc @pwittrock

**Comment by liggitt** [comment]
test would be good (both computing a patch and applying a patch to a RawExtension object)

**Comment by pwittrock** [comment]
@soltysh This is cool.

Does this relate to: https://github.com/kubernetes/community/pull/278

**Comment by pwittrock** [comment]
FYI @ymqytw and I are both out this week, but we should definitely review this.  I will try to get to it on of my evenings this week.  Plz add tests.

**Comment by soltysh** [comment]
> Does this relate to: kubernetes/community#278

It doesn't look like. It's addressing a different use case. Specifically we're having in OpenShift object that is `RawExtension` and during edit operation the patch operation fails when working against that field, instead of either dropping or replacing it (it's properly handled on the server). That's why replace is valid option for us. 

> Plz add tests.

Done, ptal.

@liggitt ptal as well

**Comment by soltysh** [comment]
Updated bazel.

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go [hunk 1]
```diff
@@ -2366,3 +2368,48 @@ func TestNumberConversion(t *testing.T) {
...
+	{
+		Description: "replace RawExtension field",
+		StrategicMergePatchRawTestCaseData: StrategicMergePatchRawTestCaseData{
+			Original: []byte(`
+replacingItem:
+  Some: Generic
+  Yaml: Inside
+  The: RawExtension
+  Field: Period
+other: current-other
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go]: include enough other fields to be able to tell that computation of other aspects of the object are still working (original field unchanged, original field modified, etc)

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go [hunk 2]
```diff
@@ -2366,3 +2368,48 @@ func TestNumberConversion(t *testing.T) {
...
+  The: RawExtension
+`),
+		},
+	},
+}
+
+func TestReplaceWithRawExtension(t *testing.T) {
+	for _, c := range replaceRawExtensionPatchTestCases {
+		original, twoWay, modified, _ := twoWayRawTestCaseToJSONOrFail(t, c)
+		actualPatch, err := CreateTwoWayMergePatch(original, modified, mergeItem)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go]: test both two and three-way

Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go]: also test applying the patch, making sure it applies successfully, and the result is what you expect (for both two and three-way)

**Comment by soltysh** [comment]
@liggitt comments addressed, ptal

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go [hunk 3]
```diff
@@ -1054,6 +1059,10 @@ func mapsHaveConflicts(typedLeft, typedRight map[string]interface{}, structType
 					return true, err
 				}
 
+				if fieldPatchStrategy == replaceDirective {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: is there a reason not to push this into mergingMapFieldsHaveConflicts? if the strategy is replace, it should return false, right?

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go [hunk 4]
```diff
@@ -158,11 +158,16 @@ func diffMaps(original, modified map[string]interface{}, t reflect.Type, ignoreC
 		switch originalValueTyped := originalValue.(type) {
 		case map[string]interface{}:
 			modifiedValueTyped := modifiedValue.(map[string]interface{})
-			fieldType, _, _, err := forkedjson.LookupPatchMetadata(t, key)
+			fieldType, fieldPatchStrategy, _, err := forkedjson.LookupPatchMetadata(t, key)
 			if err != nil {
 				return nil, err
 			}
 
+			if fieldPatchStrategy == replaceDirective {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: trying to reason through what ignoreChangesAndAdditions and ignoreDeletions would mean to this

Comment (soltysh) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: Hmmm... not sure.

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go (hunk 3 above)
Comment (soltysh) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: Actually that's reasonable, lemme try moving it up.

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go (hunk 4 above)
Comment (soltysh) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: After short investigation it looks like we should be good with the current approach of always replacing. The `ignoreChangesAndAdditions` and `ignoreDeletions` flags are used only internal in 3-way merge, which with our replace approach should be a safe call.

**Comment by soltysh** [comment]
We agreed with @liggitt that more tests around that replace mechanism will be done as a followup.

**Comment by pwittrock** [comment]
@liggit @soltysh I am out this week.  Is this urgent, or can it wait until next week?

**Comment by liggitt** [comment]
this LGTM, but I'm ok holding for @pwittrock's review

**Comment by soltysh** [comment]
@k8s-bot cvm gce e2e test this

**Comment by soltysh** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by soltysh** [comment]
@pwittrock yes it can wait until you're back, I needed Jordan's approval for the downstream work to continue

**Comment by pwittrock** [comment]
I guess this requires root level ownership?

**Comment by pwittrock** [comment]
@lavalamp WDYT of creating a second OWNERS file for strategicpatch so that we don't need root level approval?

**Comment by pwittrock** [comment]
@k8s-bot cvm gce e2e test this

**Comment by liggitt** [comment]
it's probably the vendor/BUILD change that bumped to root

**Pull request merged**